In [16]:
import xxhash

In [17]:
n = 13

In [18]:
keys = ["welcome", "to", "the", "jungle"]

In [19]:
positions = [ xxhash.xxh3_64_intdigest(k) % n for k in keys ]

In [20]:
positions

[7, 9, 10, 0]

In [21]:
keys = ["welcome", "to", "the", "jungle", "we", "got", "fun", "and", "games"]

In [22]:
positions = [ xxhash.xxh3_64_intdigest(k) % n for k in keys ]

In [23]:
positions

[7, 9, 10, 0, 2, 7, 7, 0, 11]

In [24]:
lyrics = """
Welcome to the jungle
We got fun and games
We got everything you want
Honey, we know the names
We are the people that can find
Whatever you may need
If you got no money, honey
We got your disease
In the jungle, welcome to the jungle
Watch it bring you to your knees, knees
Ooh-ah, I wanna watch you bleed
Welcome to the jungle
We take it day by day
If you want it, you're gonna bleed
But that's the price you pay
You're a very sexy girl
Who's very hard to please
You can taste the bright lights
But you won't get them for free
In the jungle, welcome to the jungle
Feel my, my, my, my serpentine
I, I wanna hear you scream
Welcome to the jungle
It's worse here everyday
You learn to live like an animal
In the jungle where we play
You got a hunger for what you see
You'll take it eventually
You can have anything you want
But you better not take it from me
In the jungle, welcome to the jungle
Watch it bring you to your knees, knees
Ooh-ah, I'm gonna watch you bleed
And when you're high, you never
Ever want to come down
So down, so down, so down, yeah
Aw
You know where you are?
You're in the jungle, baby
You're gonna die
In the jungle, welcome to the jungle
Watch it bring you to your knees, knees
In the jungle, welcome to the jungle
Feel my, my, my, my serpentine
In the jungle, welcome to the jungle
Watch it bring you to your knees, knees
In the jungle, welcome to the jungle
Watch it bring to your...
It's gonna bring you down, huh
"""

Get the set of all words in the lyrics

In [39]:
words = [ x.strip('\"') for l in lyrics.split('\n') for x in l.rstrip().split(' ')]

In [40]:
len(set(words))

119

We want a hash table large enough to accomodate all the words, we choose a prime > 119

In [50]:
n = 197

In [51]:
hashes = [xxhash.xxh3_64_intdigest(w) for w in words]

In [52]:
len(set(hashes))

119

In [53]:
hashes_modn = [h % n for h in hashes]

In [55]:
len(set(hashes_modn))

94

So we can already see that there are no collision in 64-bit hash space, but some collisions when we take the hash % n.